### ISSUE TITLE
Allow status icons on macOS to adapt to light / dark mode

### ISSUE THREAD

**Issue by samschott** [beeware/toga#2861]
### What is the problem or limitation you are having?

Menu bar icons on macOS should adapt with dark / light mode. This fits in with menu bar icons provided by the OS and most macOS apps. It also allows developers to use a black or white menu bar icon which would otherwise be almost invisible in dark or light mode, respectively. 

### Describe the solution you'd like

On macOS, icons automatically adapt to their surroundings when setting `template = true` on the NSImage. See https://developer.apple.com/documentation/appkit/nsimage/1520017-template.

We should make `template: bool` a general argument to `toga.Icon` for all platforms that support it. This would allow icons to be used more flexibly as templates and fit in more naturally in many environments besides the menu bar.

### Describe alternatives you've considered

We always set this flag in the toga-cocoa StatusIcon implementation. While macOS does allow colored menu bar icons, those are very uncommon.

**Comment by freakboy3742** [comment]
I don't know if `template=True` is the right "spelling" for this; but I agree that there's something worth capturing here - the idea of a "simple bitmap" icon, that can be masked both dark and light. 

One of the bigger questions I have is how this would be interpreted on GTK and Windows. Those platforms don't have the `template=True` argument... so we'd need to work out what a "template" icon means on those platforms. I guess we could do a "convert to 1 bit mask" and then convert to a white/black based on that mask...

**Comment by samschott** [comment]
What is toga's and your approach to APIs that might result in different results depending on the platform? For example, a `symbolic` argument that attempts to convert or declare the icon to be symbolic, but logs a warning if not supported by the platform or the provided file format?

**Comment by samschott** [comment]
Agreed on distinguishing the concepts of "mask" and "symbolic" icons, and your suggested fallbacks when not natively supported by a platform.

> mask icons are definitely a quirk of macOS operation

I believe the concept also exists on Gtk and Windows, that's what I was trying to say in https://github.com/beeware/toga/issues/2861#issuecomment-2362245024, though the nomenclature varies for the platforms.

What you call mask, Gnome seems to call symbolic (which is a different concept there from default icons for "home" etc). But it would indeed require either a correctly crafted SVG input, or some preceding conversion steps by toga.

Similarly, Windows always tries to adapt icon colors to the theme, for all of its icon classes except for [ImageIcon](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.controls.imageicon). E.g., the docs for a [BitmapIcon](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.controls.bitmapicon?view=windows-app-sdk-1.6) specify:

> The file that you use should be a solid image on a transparent background. [...]
> 
> All color info is stripped from the bitmap when the BitmapIcon is rendered. The remaining non-transparent colors are combined to produce an image that's entirely the foreground color as set by the [Foreground](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.controls.iconelement.foreground?view=windows-app-sdk-1.6#microsoft-ui-xaml-controls-iconelement-foreground) property (this typically comes from styles or templates, such as the default template resolving to a theme resource). You can override this behavior by setting the [ShowAsMonochrome](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.controls.bitmapicon.showasmonochrome?view=windows-app-sdk-1.6#microsoft-ui-xaml-controls-bitmapicon-showasmonochrome) property.

**Comment by freakboy3742** [comment]
> What you call mask, Gnome seems to call symbolic (which is a different concept there from default icons for "home" etc). But it would indeed require either a correctly crafted SVG input, or some preceding conversion steps by toga.

My apologies - I didn't follow the link you provided, assuming I knew which part of the GTK API you were referring to. I assumed you were talking about [named icons in icon themes](https://docs.gtk.org/gtk3/method.IconTheme.lookup_icon.html); I see now that symbolic icons are a whole other thing.

I'm not sure how we'd reconcile a hard SVG icon requirement, though. The best thought I've got off the top of my head would be make the list of accepted icon input formats tied to this new flag we're talking about - so a "mask" icon on GTK would *require* SVG format. This is in addition to baseline addition of support for SVG icons, which look like they will need a different loading mechanism to bitmap-based image sources.

> Similarly, Windows always tries to adapt icon colors to the theme, for all of its icon classes except for [ImageIcon](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.controls.imageicon). E.g., the docs for a [BitmapIcon](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.controls.bitmapicon?view=windows-app-sdk-1.6) specify:
> 
> > The file that you use should be a solid image on a transparent background. [...]

I'd be interested to see how this interacts with Toga's existing icon support - we're doing a straightforward "icon from file"; I presume that means we're implicitly creating Bitmap icons, as I haven't seen any color adaptation. However, the docs you've referenced here are also WinUI3 docs, not Winforms, so maybe the treatment is different due to that,

**Comment by hyuri** [comment]
Light/Dark mode support for all icons would be very useful. Especially if each icon can be single SVG app developers provide — that then gets appropriately re-colored & rendered behind-the-scenes.

Since, unlike text, icons currently do not auto adapt to light/dark mode, it ends up not making sense to use it in many cases now:

<img width="84" alt="Image" src="https://github.com/user-attachments/assets/7cbe65fb-f8d3-4f43-852e-2cf5527f2c79" />

<img width="100" alt="Image" src="https://github.com/user-attachments/assets/cb2e796b-d32a-4c67-895c-5029a3aaa00e" />
